include/deal.II/lac/solver_gmres.h

```diff
@@ -1963,39 +2128,126 @@ double SolverGMRES<VectorType>::criterion()
-SolverFGMRES<VectorType>::SolverFGMRES(SolverControl            &cn,
-                                       VectorMemory<VectorType> &mem,
-                                       const AdditionalData     &data)
+SolverMPGMRES<VectorType>::SolverMPGMRES(SolverControl            &cn,
+                                         VectorMemory<VectorType> &mem,
+                                         const AdditionalData     &data)
   : SolverBase<VectorType>(cn, mem)
   , additional_data(data)
-{}
+{
+  if (data.use_truncated_mpgmres_strategy)
+    indexing_strategy = IndexingStrategy::truncated_mpgmres;
+  else
+    indexing_strategy = IndexingStrategy::full_mpgmres;
+}
 
 
 
 template <typename VectorType>
 DEAL_II_CXX20_REQUIRES(concepts::is_vector_space_vector<VectorType>)
-SolverFGMRES<VectorType>::SolverFGMRES(SolverControl        &cn,
-                                       const AdditionalData &data)
+SolverMPGMRES<VectorType>::SolverMPGMRES(SolverControl        &cn,
+                                         const AdditionalData &data)
   : SolverBase<VectorType>(cn)
   , additional_data(data)
-{}
+{
+  if (data.use_truncated_mpgmres_strategy)
+    indexing_strategy = IndexingStrategy::truncated_mpgmres;
+  else
+    indexing_strategy = IndexingStrategy::full_mpgmres;
+}
 
 
 
 template <typename VectorType>
 DEAL_II_CXX20_REQUIRES(concepts::is_vector_space_vector<VectorType>)
-template <typename MatrixType, typename PreconditionerType>
+template <typename MatrixType, typename... PreconditionerTypes>
 DEAL_II_CXX20_REQUIRES(
   (concepts::is_linear_operator_on<MatrixType, VectorType> &&
-   concepts::is_linear_operator_on<PreconditionerType, VectorType>))
-void SolverFGMRES<VectorType>::solve(const MatrixType         &A,
-                                     VectorType               &x,
-                                     const VectorType         &b,
-                                     const PreconditionerType &preconditioner)
+   (concepts::is_linear_operator_on<PreconditionerTypes, VectorType> && ...)))
+void SolverMPGMRES<VectorType>::solve(
+  const MatrixType &A,
+  VectorType       &x,
+  const VectorType &b,
+  const PreconditionerTypes &...preconditioners)
 {
-  LogStream::Prefix prefix("FGMRES");
+  LogStream::Prefix prefix("MPGMRES");
+  SolverMPGMRES<VectorType>::solve_internal(A, x, b, preconditioners...);
+}
+
+
+
+template <typename VectorType>
+DEAL_II_CXX20_REQUIRES(concepts::is_vector_space_vector<VectorType>)
+template <typename MatrixType, typename... PreconditionerTypes>
+void SolverMPGMRES<VectorType>::solve_internal(
+  const MatrixType &A,
+  VectorType       &x,
+  const VectorType &b,
+  const PreconditionerTypes &...preconditioners)
+{
+  constexpr std::size_t n_preconditioners = sizeof...(PreconditionerTypes);
+
+  //
+  // With each invocation of preconditioner_vmult we increase the
+  // variable current_index by one modulo n_preconditioners.
+  //
```

Comment:
This is more a question of style: We do usually not use empty commented lines before and after a comment (unless in very rare occasions of long functions), so I'd prefer if we could change this here and below to
```suggestion
  // With each invocation of preconditioner_vmult we increase the
  // variable current_index by one modulo n_preconditioners.
```
(This is not an important comment, though.)

Reply:
You just haven't seen the light of proper commenting styles yet.
(I have updated the comments as you suggested.)